src/mapreduce.jl

```diff
@@ -285,3 +285,43 @@ end
         @inbounds return similar_type(a, T, Size($Snew))(tuple($(exprs...)))
     end
 end
+
+struct _InitialValue end
+
+_maybeval(dims::Integer) = Val(Int(dims))
+_maybeval(dims) = dims
+_valof(::Val{D}) where D = D
+
+@inline Base.accumulate(op::F, a::StaticVector; dims = :, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline Base.accumulate(op::F, a::StaticArray; dims, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline function _accumulate(op::F, a::StaticArray, dims::Union{Val,Colon}, init) where {F}
+    # Adjoin the initial value to `op`:
+    rf(x, y) = x isa _InitialValue ? y : op(x, y)
+
+    # StaticArrays' `reduce` is `foldl`:
+    results = _reduce(
+        a,
+        dims,
+        (init = (similar_type(a, Union{}, Size(0))(), init),),
```

Comment:
> One thing I noticed here is that for length-0 input, the result has eltype of `Union{}`.

I don't think it's possible to solve this in general as the output value depends on the function `op`.  I think another (somewhat) sensible output is to use the `eltype` of input array.  But this is not always the valid answer; e.g., `accumulate(//, [1, 2])` (note that `Base`'s `accumulate(//, Int[])` magically returns `Rational{Int64}[]` as it uses `promote_op`).

As all static arrays (including `MArray`) are shape-immutable, I don't think it is so harmful to return `Union{}` element type.

My preference is:

1. `Union{}`
2. `promote_op(op, eltype(input), eltype(input))`
3. `eltype(input)`

Reply:
I don't think it's possible to solve this in general as the output value depends on the function `op`.

This is what #664 hacks around for `map` by using `return_type` :grimacing: (but it's no worse than Base which does the same thing for `broadcast` and `collect` on empty containers). I've never been quite sure that's the right choice, but if it can't be made to work or is otherwise terrible I'm ok with `Union{}`. Prior to #664 we used `Union{}` for `map` and people mostly seemed ok with it.

Reply:
Ah, I see. I misread the last half of your first comment.

But I think one difficulty of `accumulate` compared to `map` is that output "`eltype`" is kind of ill-defined (or maybe it's better to say length-dependent).  It's conceivable to have "type sequence" like this

```julia
op(::A, ::A) :: B
op(::B, ::A) :: C
op(::C, ::A) :: D
op(::D, ::A) :: D  # finally "stabilized"
```

But, even though `B`, `C`, `D` are all different types, `promote_type(B, C, D)` may return type `E` that is a concrete type.  It is also easy to have "oscillatory" output:

```julia
osc_op(::Nothing, ::Any) = missing
osc_op(::Missing, ::Any) = nothing
```

So, maybe we should use something like this?

```julia
function f(op, acc, x)
    T = typeof(acc)
    while true
        acc = op(acc, x)
        T = promote_type(T, typeof(acc))
        non_existing_value && return T
        # `non_existing_value` modeling truncation arbitrary-sized input.
    end
end

return_type(f, Tuple{typeof(op), typeof(init), eltype(a)})
```

Impressively, `julia` can figure out something non-trivial examples like `Base._return_type(f, Tuple{typeof(+), Int, Union{Int,Missing}})` and `Base._return_type(f, Tuple{typeof(osc_op), Missing, Int})`.

I'm not sure if we need to go this far, though. `Base` is not doing this for `accumulate`.

Reply:
You're right, it's confusing to know what the empty case should do here and it's worse than map. With your fix we have

```julia
julia> accumulate((a,b)->(a+b)/2, SA[1,2])
2-element SArray{Tuple{2},Float64,1,2} with indices SOneTo(2):
 1.0
 1.5

julia> accumulate((a,b)->(a+b)/2, SA[1])
1-element SArray{Tuple{1},Int64,1,1} with indices SOneTo(1):
 1

julia> accumulate((a,b)->(a+b)/2, SA{Int}[])
0-element SArray{Tuple{0},Int64,1,0} with indices SOneTo(0)
```

It seems like we must make a somewhat arbitrary choice for the length-0 case with `init`. You've got

```julia
julia> accumulate((a,b)->(a+b)/2, SA{Int}[1], init=0)
1-element SArray{Tuple{1},Float64,1,1} with indices SOneTo(1):
 0.5

julia> accumulate((a,b)->(a+b)/2, SA{Int}[], init=0)
0-element SArray{Tuple{0},Float64,1,0} with indices SOneTo(0)
```

Overall do you feel like this is an improvement on using `Union`? I do like that `cumsum` now preserves numeric eltypes and I feel like that could be helpful for users for the same reason that "fixing" map was.

We do have oddities like

```julia
julia> cumsum(Int8[1])
1-element Array{Int64,1}:
 1

julia> cumsum(SA{Int8}[1])
1-element SArray{Tuple{1},Int8,1,1} with indices SOneTo(1):
 1

julia> cumsum(SA{Int8}[1,2])
2-element SArray{Tuple{2},Int64,1,2} with indices SOneTo(2):
 1
 3
```

I think that inconsistency between here and Base might be harmless though, and it's certainly not clear how to "fix" it.

Reply:
> Overall do you feel like this is an improvement on using `Union`?

That's a tough question... :sweat_smile:  Personally, I think we should be using `Union{}` eltype everywhere when it's not possible to figure out `eltype` without relying on the compiler. However, we don't have a good toolset and coding convention for dealing with `Union{}` eltype ("mutate-or-widen" style, even at the surface API level).

But I think the consistency of the API within a package and with respect to `Base` matters. As `map` is already using `return_type`, overall, I think using `return_type` seems to be a decent solution here.